### PR TITLE
[OCPQE-18597]Config registry storage in disconnect nutanix profile

### PR DESCRIPTION
**Pull request by wewang58**

### CODE REVIEW

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28/1739935706635046912/artifacts/nutanix-ipi-disconnected-f28

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28

**Comment by jianzhangbjz** [comment]
The `dockercfg-xxx` secrets were generated successfully.
```console
MacBook-Pro:~ jianzhang$ omg get secret -n openshift-operator-lifecycle-manager 
2024-01-03 14:45:13.773 | WARNING  | omg.utils.load_yaml:<module>:10 - yaml.CSafeLoader failed to load, using SafeLoader
NAME                                         TYPE                                 DATA  AGE
builder-dockercfg-lv927                      kubernetes.io/dockercfg              1     37m
builder-token-gkf9x                          kubernetes.io/service-account-token  4     38m
catalog-operator-serving-cert                kubernetes.io/tls                    2     58m
collect-profiles-dockercfg-xlbld             kubernetes.io/dockercfg              1     37m
collect-profiles-token-j95n6                 kubernetes.io/service-account-token  4     38m
default-dockercfg-98w96                      kubernetes.io/dockercfg              1     37m
default-token-zb6kp                          kubernetes.io/service-account-token  4     38m
deployer-dockercfg-tq96f                     kubernetes.io/dockercfg              1     37m
deployer-token-jfwbj                         kubernetes.io/service-account-token  4     38m
olm-operator-serviceaccount-dockercfg-zkp5j  kubernetes.io/dockercfg              1     37m
olm-operator-serviceaccount-token-dhll2      kubernetes.io/service-account-token  4     38m
olm-operator-serving-cert                    kubernetes.io/tls                    2     58m
package-server-manager-serving-cert          kubernetes.io/tls                    2     58m
packageserver-service-cert                   kubernetes.io/tls                    3     57m
pprof-cert                                   kubernetes.io/tls                    2     1h2m
```
But, the IR is not in a good status.
```console
MacBook-Pro:~ jianzhang$ omg get pods -n openshift-image-registry 
NAME                                              READY  STATUS   RESTARTS  AGE
cluster-image-registry-operator-574fcf54d5-l9cwx  1/1    Running  0         59m
image-registry-64669cd9c5-lhs69                   0/1    Pending  0         37m
node-ca-4d596                                     1/1    Running  0         47m
node-ca-88pf9                                     1/1    Running  0         47m
node-ca-jbtlk                                     1/1    Running  0         47m
node-ca-mkc9k                                     1/1    Running  0         47m
node-ca-mxldx                                     1/1    Running  0         47m
node-ca-s2lkk                                     1/1    Running  0         47m
MacBook-Pro:~ jianzhang$ omg get co image-registry  -o yaml
...
status:
  conditions:
  - lastTransitionTime: '2024-01-03T03:34:51Z'
    message: 'Available: The deployment does not have available replicas

      NodeCADaemonAvailable: The daemon set node-ca has available replicas

      ImagePrunerAvailable: Pruner CronJob has been created'
    reason: NoReplicasAvailable
    status: 'False'
    type: Available
  - lastTransitionTime: '2024-01-03T03:34:51Z'
    message: 'Progressing: The deployment has not completed

      NodeCADaemonProgressing: The daemon set node-ca is deployed'
    reason: DeploymentNotCompleted
    status: 'True'
    type: Progressing
  - lastTransitionTime: '2024-01-03T03:36:37Z'
    message: 'Degraded: The deployment does not have available replicas'
    reason: Unavailable
    status: 'True'
    type: Degraded

MacBook-Pro:~ jianzhang$ omg get pods image-registry-64669cd9c5-lhs69 -o yaml
...
...
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: '2024-01-03T03:35:36Z'
    message: '0/6 nodes are available: pod has unbound immediate PersistentVolumeClaims.
      preemption: 0/6 nodes are available: 6 Preemption is not helpful for scheduling..'
    reason: Unschedulable

```

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28

**Comment by jianzhangbjz** [comment]
Seems like no `pv` in this cluster.
```console
MacBook-Pro:~ jianzhang$ omg get pods 
NAME                                             READY  STATUS   RESTARTS  AGE
cluster-image-registry-operator-567bbcbb9-zxbtf  1/1    Running  0         57m
image-registry-5f57f4fc46-2676j                  0/1    Pending  0         36m
node-ca-f2p9q                                    1/1    Running  0         48m
node-ca-s798z                                    1/1    Running  0         50m
node-ca-tz66c                                    1/1    Running  0         50m
node-ca-w7qc2                                    1/1    Running  0         48m
node-ca-wcjb8                                    1/1    Running  0         50m
node-ca-xmpld                                    1/1    Running  0         48m

  conditions:
  - lastProbeTime: null
    lastTransitionTime: '2024-01-03T13:24:38Z'
    message: '0/6 nodes are available: pod has unbound immediate PersistentVolumeClaims.
      preemption: 0/6 nodes are available: 6 Preemption is not helpful for scheduling..'
    reason: Unschedulable
    status: 'False'
    type: PodScheduled

MacBook-Pro:~ jianzhang$ omg get pv
No resources found
MacBook-Pro:~ jianzhang$ omg get pvc
2024-01-04 09:35:13.649 | WARNING  | omg.utils.load_yaml:<module>:10 - yaml.CSafeLoader failed to load, using SafeLoader
NAME                    STATUS   VOLUME  CAPACITY  ACCESS MODES  STORAGECLASS  AGE
image-registry-storage  Pending          Unkown    RWO                         37m
MacBook-Pro:~ jianzhang$ omg get pvc image-registry-storage -o yaml
2024-01-04 09:35:25.021 | WARNING  | omg.utils.load_yaml:<module>:10 - yaml.CSafeLoader failed to load, using SafeLoader
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  creationTimestamp: '2024-01-03T13:23:53Z'
  finalizers:
  - kubernetes.io/pvc-protection
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:spec:
        f:accessModes: {}
        f:resources:
          f:requests:
            .: {}
            f:storage: {}
        f:volumeMode: {}
    manager: kubectl-create
    operation: Update
    time: '2024-01-03T13:23:53Z'
  name: image-registry-storage
  namespace: openshift-image-registry
  resourceVersion: '35428'
  uid: a7cc21aa-42b3-4469-b843-6ff4e17fca88
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 100Gi
  volumeMode: Filesystem
status:
  phase: Pending

```

**Comment by wewang58** [comment]
> Seems like no `pv` in this cluster.
> 
> ```
> MacBook-Pro:~ jianzhang$ omg get pods 
> NAME                                             READY  STATUS   RESTARTS  AGE
> cluster-image-registry-operator-567bbcbb9-zxbtf  1/1    Running  0         57m
> image-registry-5f57f4fc46-2676j                  0/1    Pending  0         36m
> node-ca-f2p9q                                    1/1    Running  0         48m
> node-ca-s798z                                    1/1    Running  0         50m
> node-ca-tz66c                                    1/1    Running  0         50m
> node-ca-w7qc2                                    1/1    Running  0         48m
> node-ca-wcjb8                                    1/1    Running  0         50m
> node-ca-xmpld                                    1/1    Running  0         48m
> 
>   conditions:
>   - lastProbeTime: null
>     lastTransitionTime: '2024-01-03T13:24:38Z'
>     message: '0/6 nodes are available: pod has unbound immediate PersistentVolumeClaims.
>       preemption: 0/6 nodes are available: 6 Preemption is not helpful for scheduling..'
>     reason: Unschedulable
>     status: 'False'
>     type: PodScheduled
> 
> MacBook-Pro:~ jianzhang$ omg get pv
> No resources found
> MacBook-Pro:~ jianzhang$ omg get pvc
> 2024-01-04 09:35:13.649 | WARNING  | omg.utils.load_yaml:<module>:10 - yaml.CSafeLoader failed to load, using SafeLoader
> NAME                    STATUS   VOLUME  CAPACITY  ACCESS MODES  STORAGECLASS  AGE
> image-registry-storage  Pending          Unkown    RWO                         37m
> MacBook-Pro:~ jianzhang$ omg get pvc image-registry-storage -o yaml
> 2024-01-04 09:35:25.021 | WARNING  | omg.utils.load_yaml:<module>:10 - yaml.CSafeLoader failed to load, using SafeLoader
> apiVersion: v1
> kind: PersistentVolumeClaim
> metadata:
>   creationTimestamp: '2024-01-03T13:23:53Z'
>   finalizers:
>   - kubernetes.io/pvc-protection
>   managedFields:
>   - apiVersion: v1
>     fieldsType: FieldsV1
>     fieldsV1:
>       f:spec:
>         f:accessModes: {}
>         f:resources:
>           f:requests:
>             .: {}
>             f:storage: {}
>         f:volumeMode: {}
>     manager: kubectl-create
>     operation: Update
>     time: '2024-01-03T13:23:53Z'
>   name: image-registry-storage
>   namespace: openshift-image-registry
>   resourceVersion: '35428'
>   uid: a7cc21aa-42b3-4469-b843-6ff4e17fca88
> spec:
>   accessModes:
>   - ReadWriteOnce
>   resources:
>     requests:
>       storage: 100Gi
>   volumeMode: Filesystem
> status:
>   phase: Pending
> ```

@jianzhangbjz  I will talk with @wduan to discuss the issue, i can also reproduce the issue in jenkins

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28

**Comment by wewang58** [comment]
@jianzhangbjz  when use chain:[ ipi-install-nutanix](https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/install/nutanix/ipi-install-nutanix-chain.yaml) to install csi and configure registry with pvc in disconnected env, will met failed to pull image:[quay.io/ntnx-csi/nutanix-csi-operator-catalog:latest](https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/install/nutanix/csi/ipi-install-nutanix-csi-commands.sh#L28C10-L28C62) in the step, do you know how to resolve the issue, or  let's configure image registry storage with Emptydir{}

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28

**Comment by jianzhangbjz** [comment]
Hi @wewang58 , I guess you mean this step error: [ipi-install-nutanix-csi/build-log.txt ](https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/47159/rehearse-47159-periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28/1742807800431513600/artifacts/nutanix-ipi-disconnected-f28/ipi-install-nutanix-csi/build-log.txt) 

```console
MacBook-Pro:step-registry jianzhang$ omg get pods  -n openshift-marketplace nutanix-csi-operator-beta-7wm8z -o yaml
apiVersion: v1
kind: Pod
metadata:
...
  containerStatuses:
  - image: quay.io/ntnx-csi/nutanix-csi-operator-catalog:latest
    imageID: ''
    lastState: {}
    name: registry-server
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        message: Back-off pulling image "quay.io/ntnx-csi/nutanix-csi-operator-catalog:latest"
        reason: ImagePullBackOff
  hostIP: 10.0.133.164
  phase: Pending
  podIP: 10.130.2.14
```
And, I don't think this [ipi-install-nutanix](https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/install/nutanix/ipi-install-nutanix-chain.yaml) chain or [ipi-install-nutanix-csi](https://github.com/openshift/release/tree/master/ci-operator/step-registry/ipi/install/nutanix/csi) step is appropriate for the disconnected env. + @sgaoshang and @duanwei33 , do you happen to know which chian or csi step is appropriate for the disconnected Nutanix env? Thanks!

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28

**Comment by wewang58** [comment]
> Hi @wewang58 , I guess you mean this step error: [ipi-install-nutanix-csi/build-log.txt ](https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/47159/rehearse-47159-periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28/1742807800431513600/artifacts/nutanix-ipi-disconnected-f28/ipi-install-nutanix-csi/build-log.txt)
> 
> ```
> MacBook-Pro:step-registry jianzhang$ omg get pods  -n openshift-marketplace nutanix-csi-operator-beta-7wm8z -o yaml
> apiVersion: v1
> kind: Pod
> metadata:
> ...
>   containerStatuses:
>   - image: quay.io/ntnx-csi/nutanix-csi-operator-catalog:latest
>     imageID: ''
>     lastState: {}
>     name: registry-server
>     ready: false
>     restartCount: 0
>     started: false
>     state:
>       waiting:
>         message: Back-off pulling image "quay.io/ntnx-csi/nutanix-csi-operator-catalog:latest"
>         reason: ImagePullBackOff
>   hostIP: 10.0.133.164
>   phase: Pending
>   podIP: 10.130.2.14
> ```
> 
> And, I don't think this [ipi-install-nutanix](https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/install/nutanix/ipi-install-nutanix-chain.yaml) chain or [ipi-install-nutanix-csi](https://github.com/openshift/release/tree/master/ci-operator/step-registry/ipi/install/nutanix/csi) step is appropriate for the disconnected env. + @sgaoshang and @duanwei33 , do you happen to know which chian or csi step is appropriate for the disconnected Nutanix env? Thanks!

@jianzhangbjz  I talked with @sgaoshang and @duanwei33  seems no available steps with csi storage install for disconnected cluster,  @sgaoshang  and @duanwei33  could you help to  check again, thanks, so i want to change to use ecmptydir for registry storage.

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28

**Comment by wewang58** [comment]
@jianzhangbjz @sgaoshang  from the [job](f[unctionality-testing/aos-4_15/ipi-on-ibmcloud/versioned-installer-ovn](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/47159/rehearse-47159-periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-nutanix-ipi-disconnected-f28/1743225488597323776), image registry is available and will not affect other teams cases test, i think the pr can be merged, could you review again? 
if in future there's a step about install the Nutanix CSI driver on the disconnected cluster, we can configure registry storage with pvc in future too.

**Comment by wewang58** [comment]
@jianlinliu  Could you help to review it too, thanks a lot!

File: ci-operator/step-registry/configure-registry-storage/nutanix/configure-registry-storage-nutanix-commands.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+set -o pipefail
+
+export KUBECONFIG=${SHARED_DIR}/kubeconfig
+
+echo "$(date -u --rfc-3339=seconds) - Configuring image registry with emptyDir..."
+oc patch configs.imageregistry.operator.openshift.io cluster --type merge --patch '{"spec":{"managementState":"Managed","storage":{"emptyDir":{}}}}'
+echo "$(date -u --rfc-3339=seconds) - Wait for the imageregistry operator to see that it has work to do..."
+sleep 30
+echo "$(date -u --rfc-3339=seconds) - Wait for the imageregistry operator to go available..."
+oc wait --all --for=condition=Available=True clusteroperators.config.openshift.io --timeout=10m
```
Comment (jianlinliu) [inline ci-operator/step-registry/configure-registry-storage/nutanix/configure-registry-storage-nutanix-commands.sh:14]: This is waiting for all the operators get ready?

Comment (wewang58) [reply ci-operator/step-registry/configure-registry-storage/nutanix/configure-registry-storage-nutanix-commands.sh:14]: The patch operation will also affect other co rollout, image registry is already available after patch, but other co are still in progress
image

File: ci-operator/step-registry/configure-registry-storage/nutanix/configure-registry-storage-nutanix-ref.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+ref:
+  as: configure-registry-storage-nutanix
+  from_image:
+    namespace: ocp
+    name: "4.12"
+    tag: upi-installer
```
Comment (jianlinliu) [inline ci-operator/step-registry/configure-registry-storage/nutanix/configure-registry-storage-nutanix-ref.yaml:6]: If not necessary, suggest not use upi-installer, I guess you can use `from: cli` image.

Comment (wewang58) [reply ci-operator/step-registry/configure-registry-storage/nutanix/configure-registry-storage-nutanix-ref.yaml:6]: @jianlinliu  You mean update as below, right?
```
ref:
  as: configure-registry-storage-nutanix
  from: cli
  commands: configure-registry-storage-nutanix-commands.sh
  resources:
```

Comment (jianlinliu) [reply ci-operator/step-registry/configure-registry-storage/nutanix/configure-registry-storage-nutanix-ref.yaml:6]: right

**Pull request merged**